src/services/obs/dir_stream.rs

```diff
@@ -0,0 +1,239 @@
+                let output: Output = de::from_reader(bs.reader())
+                    .map_err(|e| other(ObjectError::new("list", &self.path, e)))?;
+
+                // Try our best to check whether this list is done.
+                //
+                // - Check `next_marker`
+                if let Some(next_marker) = output.next_marker.as_ref() {
+                    self.done = next_marker.is_empty();
+                };
+                self.next_marker = output.next_marker.clone().unwrap_or_default();
+                self.state = State::Listing((output, 0, 0));
+                self.poll_next(cx)
+            }
+            State::Listing((output, common_prefixes_idx, objects_idx)) => {
+                if let Some(prefixes) = &output.common_prefixes {
+                    if *common_prefixes_idx < prefixes.len() {
+                        *common_prefixes_idx += 1;
+                        let prefix = &prefixes[*common_prefixes_idx - 1].prefix;
+
+                        let de = DirEntry::new(
+                            backend.clone(),
+                            ObjectMode::DIR,
+                            &backend.get_rel_path(prefix),
+                        );
+
+                        debug!(
+                            "dir object {} got entry, mode: {}, path: {}",
+                            &self.path,
+                            de.mode(),
+                            de.path()
+                        );
+                        return Poll::Ready(Some(Ok(de)));
+                    }
+                };
+
+                let objects = &output.contents;
+                while *objects_idx < objects.len() {
+                    let object = &objects[*objects_idx];
+                    *objects_idx += 1;
+
+                    if object.key.ends_with('/') {
+                        continue;
+                    }
+
+                    let de = DirEntry::new(
+                        backend.clone(),
+                        ObjectMode::FILE,
+                        &backend.get_rel_path(&object.key),
+                    );
+
+                    debug!(
+                        "dir object {} got entry, mode: {}, path: {}",
+                        &self.path,
+                        de.mode(),
+                        de.path(),
+                    );
+                    return Poll::Ready(Some(Ok(de)));
+                }
+
+                if self.done {
+                    debug!("dir object {} list done", &self.path);
+                    return Poll::Ready(None);
+                }
+
+                self.state = State::Idle;
+                self.poll_next(cx)
+            }
+        }
+    }
+}
+
+#[derive(Default, Debug, Deserialize)]
+#[serde(default, rename_all = "PascalCase")]
+struct Output {
+    name: String,
+    prefix: String,
+    contents: Vec<Content>,
+    common_prefixes: Option<Vec<CommonPrefix>>,
+    marker: String,
+    #[serde(rename = "NextMarker")]
```

Comment:
Since we already have `rename_all = "PascalCase"`, we don't need to specify it again?